thruster_allocation_matrix_controller/src/thruster_allocation_matrix_controller.cpp

```diff
@@ -0,0 +1,259 @@
+  reference_interfaces.reserve(reference_interfaces_.size());
+
+  for (size_t i = 0; i < dof_; ++i) {
+    reference_interfaces.emplace_back(
+      get_node()->get_name(), dof_names_[i] + "/" + hardware_interface::HW_IF_EFFORT, &reference_interfaces_[i]);
+  }
+
+  return reference_interfaces;
+}
+
+controller_interface::return_type ThrusterAllocationMatrixController::update_reference_from_subscribers(
+  const rclcpp::Time & /*time*/, const rclcpp::Duration & /*period*/)
+{
+  auto * current_reference = reference_.readFromNonRT();
+
+  // define wrench list and iterate through to assign it to interfaces
+  std::vector<double> wrench = {(*current_reference)->force.x,  (*current_reference)->force.y,
+                                (*current_reference)->force.z,  (*current_reference)->torque.x,
+                                (*current_reference)->torque.y, (*current_reference)->torque.z};
+
+  // Update the thrust reference
+  for (size_t i = 0; i < wrench.size(); i++) {
+    if (!std::isnan(wrench[i])) {
+      reference_interfaces_[i] = wrench[i];
+      wrench[i] = std::numeric_limits<double>::quiet_NaN();
+    }
+  }
+
+  return controller_interface::return_type::OK;
+}
+
+void ThrusterAllocationMatrixController::update_parameters()
+{
+  // Need this
+  if (!param_listener_->is_old(params_)) {
+    return;
+  }
+  param_listener_->refresh_dynamic_parameters();
+  params_ = param_listener_->get_params();
+}
+
+controller_interface::CallbackReturn ThrusterAllocationMatrixController::configure_parameters()
+{
+  update_parameters();
+
+  // These are just used to improve readability
+  dof_names_ = params_.dof_names;
+  dof_ = dof_names_.size();
+
+  // stroing each row of the tam matrix
+  Eigen::Map<Eigen::VectorXd> tam_x(params_.tam.x.data(), params_.tam.x.size());
+  Eigen::Map<Eigen::VectorXd> tam_y(params_.tam.y.data(), params_.tam.y.size());
+  Eigen::Map<Eigen::VectorXd> tam_z(params_.tam.z.data(), params_.tam.z.size());
+  Eigen::Map<Eigen::VectorXd> tam_rx(params_.tam.rx.data(), params_.tam.rx.size());
+  Eigen::Map<Eigen::VectorXd> tam_ry(params_.tam.ry.data(), params_.tam.ry.size());
+  Eigen::Map<Eigen::VectorXd> tam_rz(params_.tam.rz.data(), params_.tam.rz.size());
+
+  // checking that each of the vectors is the same size
+  num_thrusters_ = tam_x.size();
+  if (
+    (tam_y.size() != num_thrusters_) || (tam_z.size() != num_thrusters_) || (tam_rx.size() != num_thrusters_) ||
+    (tam_ry.size() != num_thrusters_) || (tam_rz.size() != num_thrusters_)) {
+    // if any of the tam vectors aren't the same size, then we need to notify the user and return failure
+    RCLCPP_ERROR(get_node()->get_logger(), "Received vectors for TAM with differing lengths.");
+    return controller_interface::CallbackReturn::ERROR;
+  }
+
+  // otherwise, they are all the same length and we can pack them into an Eigen matrix
+  Eigen::MatrixXd tam_(dof_, num_thrusters_);
+  tam_ << tam_x, tam_y, tam_z, tam_rx, tam_ry, tam_rz;
```

Comment:
Check this in your debugging. I think Eigen is column major so things might be inserted weird.

Reply:
Okay, will do.